### PR TITLE
Command and Missions

### PR DESCRIPTION
**Pull request by mattlangford**
Add missions for reaching Saturday's goals.

Some aliases I have for operating the boat:
```shell
alias navm='rosrun navigator_missions run_mission'
alias navc='rosrun navigator_missions cmd_mission'
```

Command line missions can be run like this:
To move forward 1 meter at a speed of .5 m/s:
```shell
navc forward 1 -s .5
````
To move right 15 feet at max speed:
```shell
navc right 15ft
````

To yaw right 15 degrees at max speed:
```shell
navc yaw_right 15deg
````

Default units are meters and radians

### CODE REVIEW

File: mission_control/navigator_missions/nav_missions/back_and_forth.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+import txros
+
+
+@txros.util.cancellableInlineCallbacks
+def main(navigator):
+    while True:
+        point_2 = navigator.move
```
Comment (DSsoto) [inline mission_control/navigator_missions/nav_missions/back_and_forth.py]: I'm assuming that navigator.move is the current position? Why?

File: mission_control/navigator_missions/navigator_singleton/navigator.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        self._change_wrench = yield self.nh.get_service_client('/change_wrench', navigator_svrs.WrenchSelect)
+        #self._enu_odom_sub = yield self.nh.subscribe('world_odom', Odometry)
+        self.tf_listener = yield tf.TransformListener(self.nh)
+
+        yield self.nh.sleep(.3)  # Build tf and odom buffers
+        defer.returnValue(self)
+
+    @property
+    def pose(self):
+        # TODO: Fails to get last message when
```
Comment (DSsoto) [inline mission_control/navigator_missions/navigator_singleton/navigator.py]: when what?

File: mission_control/navigator_missions/navigator_singleton/navigator.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        resp = yield s_client(s_req)
+        defer.returnValue(resp)
+
+    def _load_vision_serivces(self, fname="vision_services.yaml"):
+        rospack = rospkg.RosPack()
+        config_file = os.path.join(rospack.get_path('navigator_missions'), 'navigator_singleton', fname)
+        f = yaml.load(open(config_file, 'r'))
+        for name in f:
+            s_type = getattr(navigator_svrs, f[name]["type"])
```
Comment (DSsoto) [inline mission_control/navigator_missions/navigator_singleton/navigator.py]: "navigator_srvs"

File: mission_control/navigator_missions/navigator_singleton/navigator.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    nh = yield NodeHandle.from_argv("testing")
+    n = yield _Navigator(nh)._init()
+    print (yield n.vision_request('sonar'))
+
+
+if __name__ == '__main__':
+    from twisted.internet import defer, reactor
+    import signal
+    import traceback
+    reactor.callWhenRunning(main)
```
Comment (DSsoto) [inline mission_control/navigator_missions/navigator_singleton/navigator.py]: What does reactor do? Why do you need `def(main)` with the usual `if __name__ == '__main__':` already there?

File: mission_control/navigator_missions/navigator_singleton/pose_editor.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+
+UP = np.array([0.0, 0.0, 1.0], np.float64)
+EAST, NORTH, WEST, SOUTH = [transformations.quaternion_about_axis(np.pi / 2 * i, UP) for i in xrange(4)]
+UNITS = {'m': 1, 'ft': 0.3048, 'yard': 0.9144, 'rad': 1, 'deg': 0.0174533}
+
+def normalized(x):
+    x = np.array(x)
+    if max(map(abs, x)) == 0:
+        warnings.warn('Normalizing zero-length vector to random unit vector')
```
Comment (DSsoto) [inline mission_control/navigator_missions/navigator_singleton/pose_editor.py]: Don't these kind of errors ocur in several places? It would be nice to include the file name or node name in the message.

File: mission_control/navigator_missions/navigator_singleton/pose_editor.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        [(a, b, c, 0) for a, b, c in rot] +
+        [(0, 0, 0, 1)])
+
+
+def look_at(forward, upish=UP):
+    # assumes standard forward-left-up body coordinate system
+    return triad((forward, upish), ([1, 0, 0], UP))
+
+
+def look_at_without_pitching(forwardish, up=UP):
```
Comment (DSsoto) [inline mission_control/navigator_missions/navigator_singleton/pose_editor.py]: Can we even pitch? Aren't all the thrusters in a plane?

File: mission_control/navigator_missions/nav_missions/back_and_forth.py (hunk 1 above)
Comment (mattlangford) [inline mission_control/navigator_missions/nav_missions/back_and_forth.py]: navigator.move == navigator.move.forward(0)
I'll add a comment to explain it

File: mission_control/navigator_missions/navigator_singleton/navigator.py (hunk 3 above)
Comment (mattlangford) [inline mission_control/navigator_missions/navigator_singleton/navigator.py]: derp

File: mission_control/navigator_missions/navigator_singleton/navigator.py (hunk 4 above)
Comment (mattlangford) [inline mission_control/navigator_missions/navigator_singleton/navigator.py]: In this context the reactor is a way to call a decorated cancellableInlineCallbacks function. If you didn't use a reactor and just ran the `main()` function directly, you would get back a canceled deffered since the function call was not in a decorated cancellableInlineCallbacks function.

**TLDR** It's a way to call twisted functions from non twisted functions and not have them break

File: mission_control/navigator_missions/navigator_singleton/pose_editor.py (hunk 6 above)
Comment (mattlangford) [inline mission_control/navigator_missions/navigator_singleton/pose_editor.py]: These are ported from the sub, I will remove since they don't make sense in this context

**Comment by DSsoto** [comment]
What's the difference between a run mission and a command mission?
Is it documented anywhere?

**Comment by DSsoto** [comment]
Just a couple of minor comments, this is pretty good code though. I think you should make an effort to add more comments to your code, especially for things that you're doing that would not be easy to understand by your average reader. 

I can merge tonight before lake day when you're done.
